File: desktop-client/src/app/game/services/participant/participant.service.ts
import { Injectable } from '@angular/core';
import { MatSnackBar } from '@angular/material/snack-bar';
import { Router } from '@angular/router';
import { SNACK_CLOSE_ICON_ACTION } from '@app/constants/snack-bar-constants';
import { PageUrl } from '@app/enum/page-url';
import { PLAYER_BANNED_MESSAGE } from '@app/game/constants/banned-message';
import { WebsocketService } from '@app/shared/services/websocket/websocket.service';
import { PlayerDto } from '@common/interfaces/playerDto';
import { RoomEvent } from '@common/socket-events/room-event';

@Injectable({
    providedIn: 'root',
})
export class ParticipantService {
    participants: PlayerDto[] = [];

    constructor(
        private webSocketService: WebsocketService,
        private readonly router: Router,
        private readonly snackBar: MatSnackBar,
    ) {}

    initializeParticipantsListener() {
        this.setupParticipantsListListener();
        this.setupParticipantsBanListener();
        this.updateParticipants();
    }

    removeParticipantsListener() {
        this.webSocketService.removeAllListeners(RoomEvent.UPDATE_PARTICIPANT_LIST);
        this.webSocketService.removeAllListeners(RoomEvent.BAN);
    }

    removeParticipant(playerName: string) {
        this.webSocketService.send(RoomEvent.BAN, playerName);
    }

    clearParticipants() {
        this.participants = [];
    }

    private updateParticipants() {
        this.webSocketService.send(RoomEvent.UPDATE_PARTICIPANT_LIST);
    }

    private setupParticipantsListListener() {
        this.webSocketService.on(RoomEvent.UPDATE_PARTICIPANT_LIST, (participants?: PlayerDto[]) => {
            if (participants) this.participants = participants;
        });
    }

    private setupParticipantsBanListener() {
        this.webSocketService.on(RoomEvent.BAN, () => {
            this.router.navigate([`/${PageUrl.APP_PREFIX}/${PageUrl.HOME}`]);
            this.snackBar.open(PLAYER_BANNED_MESSAGE, SNACK_CLOSE_ICON_ACTION);
        });
    }
}

File: mobile_client/lib/services/participant_service.dart
import 'package:flutter/material.dart';
import 'package:mobile_client/constants/enum/page_url.dart';
import 'package:mobile_client/constants/player_banned_message.dart';
import 'package:mobile_client/constants/snack_bar_constants.dart';
import 'package:mobile_client/constants/socket_events/room_event.dart';
import 'package:mobile_client/models/player_dto.dart';
import 'package:mobile_client/services/websocket_service.dart';
import 'package:mobile_client/utils/app_navigator.dart';

class ParticipantService with ChangeNotifier {
  final WebSocketService webSocketService;
  List<PlayerDto> participants = [];

  ParticipantService(this.webSocketService);

  void initializeParticipants(BuildContext context) {
    setupParticipantsListListener();
    setupParticipantsBanListener(
        context); // TODO: this is not needed for organizer
    updateParticipants();
  }

  removeParticipantsListener() {
    webSocketService.removeAllListeners(RoomEvent.updateParticipantList.name);
    webSocketService.removeAllListeners(RoomEvent.ban.name);
  }

  void removeParticipant(String playerName) {
    webSocketService.send(RoomEvent.ban.name, playerName);
  }

  void clearParticipants() {
    participants.clear();
  }

  void updateParticipants() {
    webSocketService.send(RoomEvent.updateParticipantList.name);
  }

  void setupParticipantsListListener() {
    webSocketService.on<List<dynamic>>(RoomEvent.updateParticipantList.name,
        (data) {
      participants = data.map((item) => PlayerDto.fromJson(item)).toList();
      notifyListeners();
    });
  }

  void setupParticipantsBanListener(BuildContext context) {
    webSocketService.on(RoomEvent.ban.name, (_) {
      AppNavigator.go('/${PageUrl.home.name}');
      ScaffoldMessenger.of(context).showSnackBar(
        SnackBar(
          content: const Text(playerBannedMessage),
          duration: Duration(seconds: 3),
          behavior: SnackBarBehavior.floating,
          action: SnackBarAction(
            label: snackCloseIconAction,
            onPressed: () {},
          ),
          shape: RoundedRectangleBorder(
            borderRadius: BorderRadius.all(Radius.circular(4)),
          ),
          width: 300,
        ),
      );
    });
  }
}
